src/clusterfuzz/_internal/bot/tasks/cron/helpers/bot_manager.py

```diff
@@ -0,0 +1,318 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Library to manage bots using GCE instance templates and groups."""
+
+import time
+
+import google_auth_httplib2
+import googleapiclient
+from googleapiclient.discovery import build
+import httplib2
+
+from clusterfuzz._internal.base import retry
+from clusterfuzz._internal.google_cloud_utils import credentials
+
+RETRY_COUNT = 8
+RETRY_DELAY = 4
+
+REQUEST_TIMEOUT = 180
+
+_SCOPES = [
+    'https://www.googleapis.com/auth/cloud-platform',
+]
+
+
+class BotManagerError(Exception):
+  """Base exception class."""
+
+
+class OperationError(BotManagerError):
+  """Errors during an operation."""
+
+
+class RequestError(BotManagerError):
+  """Errors during a request."""
+
+
+class NotFoundError(RequestError):
+  """Not found."""
+
+
+class AlreadyExistsError(RequestError):
```

Comment:
It's a minor issue but I'd probably use fewer error classes and let the message be descriptive instead of making so many.